### PR TITLE
#27_account creation _page

### LINKED ISSUE #27: Add an account creation page
As a user,
I want to create an account,
So that I can keep track of my activity in the app.

- When the user visits `/signup`, they see a form with fields for Name, Email and Password
- The form includes a submit button 
- When the form is submitted, the app submits a `POST` request to `/users` on the backend with the form data included in the post body (communicate with the backend team to determine the format of the request)
- If the account creation is successful, the user is directed to their account page (/account) which displays the user's name and email address

### PR DESCRIPTION
**Pull request by Victoria240**
## Change Summary

- Add component \AccountCreation & AuthenticatedContent
- Update feature AccountCreation to include POST request from /users endpoint when Mo creates one

## I Had Difficulty With Connecting to the backend because I had no .env file 

_I initially tried to implement this using our existing backend files but ran into a "cannot find dotenv module" error so I decided to install env for this task._

### CODE REVIEW

File: tsconfig.json
Comment (dargaCode) [inline tsconfig.json]: Please revert this line, I believe it's what's causing the "module not found" error from #36

File: package-lock.json
Comment (dargaCode) [inline package-lock.json:1]: You're not adding any new libraries, so this file shouldn't be changing at all. Please delete the file then `npm install`, then hopefully it will show no changes

File: src/components/AuthenticatedContent/AuthenticatedContent.tsx
Comment (dargaCode) [inline src/components/AuthenticatedContent/AuthenticatedContent.tsx]: We can remove the `password` entry, since we won't ever want to show that to the user

File: src/components/AuthenticatedContent/AuthenticatedContent.tsx
Comment (dargaCode) [inline src/components/AuthenticatedContent/AuthenticatedContent.tsx:1]: Awesome!

File: src/App.tsx
Comment (dargaCode) [inline src/App.tsx]: We can remove `password` from here too

```suggestion
                <AuthenticatedContent username={""} email={""} />
```

File: src/App.tsx
Comment (dargaCode) [inline src/App.tsx]: nice!

Let's find out from backend what type of authentication data we can expect to receive from them, so we can make sure these checks will work with the api 👍

File: src/App.tsx
Comment (dargaCode) [inline src/App.tsx]: my ESlint is saying that these imports aren't sorted correctly. I made myself a bug #43 to make sure ESlint is run on the code from every commit. 

In the meantime though, please run `npm run lint` locally to fix the sort order. 

We should also get your editor set up so that it runs Prettier and ESlint every time you hit save on a file

File: src/components/AccountCreation/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreation.tsx]: My chrome console is pointing out that password inputs should always be wrapped in a `form` element:

Screen Shot 2023-11-29 at 5 19 34 PM

If you replace this `div` with a form`, you can remove the event handler that's currently the `onClick` for the submit button, and move it to the form's `onSubmit` event handler: https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit_event

File: src/components/AccountCreation/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreation.tsx]: later on, we'll be able to add some fancy things to the event handlers on these inputs, like checking for a valid email address, and validating that the username or email isn't already taken 🙌

File: src/App.tsx
Comment (dargaCode) [inline src/App.tsx]: Really great job in this PR overall with using `on-` vs `handle-` prefixes accurately for the event handlers!

File: src/components/AccountCreation/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreation.tsx]: Great job on the props here! There are a couple of different ways to present this kind of thing, including this alternate way: 

```suggestion
interface AccountCreationProps {
  onAccountCreate: (username: string, email: string, password: string) => void;
}

function AccountCreation(props:  AccountCreationProps): JSX.Element {
  const {onAccountCreate} = props;
```

In the meantime though, we said we're going to stick to vanilla JS for the new components we create for now. Can you please remove the types from these components, and add the comments at the top to disable typescript checking?

```
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
```

**Review by dargaCode — CHANGES_REQUESTED** [review]
Great job Victoria! Please remove the typescript stuff for the time being since we said we'll stick to vanilla js going forward 🙏

File: tsconfig.json
Comment (Victoria240) [reply tsconfig.json]: I pulled the changes in the dev branch and can't see the error anymore

File: src/components/HomePage/HomePage.tsx
Comment (dargaCode) [inline src/components/HomePage/HomePage.tsx:1]: Please remove this empty file for now 🙏

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: Very small nitpick, but please put these disabling comments at the very top of the file:

```suggestion
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import React, { useState } from "react";

```

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: `// @ts-ignore` disables type checking  for the whole file, so you only need these lines once at the top

```suggestion
```

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx:1]: We only need the `.tsx` extension at the end of the filename itself. It's not breaking anything to have it in the folder name too, but it's just kindof misleading and confusing (especially since more than one file will end up in this folder)

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: Disallowing strings of only spaces is a great catch! 💪

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: We should avoid using `<br />` in favor of wrapping elements in `<div>` when we want to force a line break. 

The reason this is happening is that `<label>` is a `display: inline` element,by default and the reason `<div>` fixes it is that it's a `display: block` element by default

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: please replace with `<div>` wrapper

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: please replace with `<div>` wrapper

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: The backend validation is the most important, but HTML also has some built in input validation we can use for now to give people clearer messaging when they've got invalid data: 

`required` is a shortcut for `required={true}`; same works with other boolean attributes like `disabled`

```suggestion
          <input
            required
            type="text"
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
```

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: In addition to making the password required, we can also require the password to have a minimum length:

(According to the backend error message, 8 is the minimum password length required)

```suggestion
          <input
            required
            type="password"
            value={password}
            minLength={8}
            onChange={(e) => setPassword(e.target.value)}
          />
```

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation.tsx/AccountCreation.tsx]: All 3 of these `<Input>`s can be marked as required to match what the backend considers them

```suggestion
          <input
            required
            type="text"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
          />
```

**Review by dargaCode — CHANGES_REQUESTED** [review]
Nice work Victoria, thank you! 

This code review will be a lot to take in, so let's please get on a call and go over it together to make sure it's all feeling ok 🤝 

I've got some other small changes here and then this one larger file structure change, which you should tackle last:

There are some best practices we can follow for how our components are broken out and what they contain, which I'm sorry I didn't think to talk about before we started you on this task! 

Usually the `<App>` component should generally just handle the routing, but not have any state or handlers defined in it, or other conditional rendering within those routes. 

I don't know if you remember [the container/view pattern ](https://reactpatterns.com/#container-component)that we looked at in the todo list.

For things to be the easiest to update over time, it's best if we can count on all the same related code to be all together in the same component, particularly that we don't touch `<App>` very much for feature stuff, and for a given feature we also look at a two different components depending on whether we want to see API fetching code or UI rendering code. 

For this feature it would break down into these 3 files:
- `<App>` would do only the routing
- `<AccountCreationContainer>` would handle all the he API interactions and Auth
- `<AccountCreation>` would only render the sign up form U
- `<AuthenticatedContent>` would render all the actual content of the app

(when we get further into Auth stuff this will probably need updating a bit too, but tbh I have to go learn how to do that 😅)

From where we are now, we can get to this goal by:
1. pulling all of our API interactions and API requests for each feature into a new "container" component, and 
2. leaving only the rendering in our "presentational components" (the already-existing`<AccountCreation> ` and `<AuthenticatedContent>` components)

So we want to end up in a file structure where:
* `<App />` renders new component `<AcccountCreationContainer>` in the `<Route>`.
* `<AcccountCreationContainer >` does:
      - all fetching/posting of  Users
      - `isAuthenticated` state lives here
      -  the conditional rendering for `<AccountCreation>` vs `<AuthenticatedContent>` lives here too.
- Then `<AccountCreation>` receives `onCreateAccount` as a prop, (and does no fetching itself)
- And `<AuthenticatedContent>` receives the same user data as it does now.

File: src/components/AccountCreation.tsx/AccountCreation.tsx
Comment (dargaCode) [reply src/components/AccountCreation.tsx/AccountCreation.tsx]: Sorry I was wrong, we should be using `@ts-nocheck` instead of `@ts-ignore`!

**Review by msrezaie — COMMENTED** [review]
Hey Victoria, great job on adding the account creation page. 🎉You've mentioned a POST request '/users' endpoint, by this endpoint do you mean for registering/creating a user? If so, the endpoint for this is already implemented. If you send a POST request to the following endpoint:

/api/v1/auth/signup 

with a request body of:
{
  "name": "victoria",
  "email": "victoria@email.com",
  "password": "aStrongPassword123"
}

A new user with the given detail will be created and signed in.

**Comment by Victoria240** [comment]
> Hey Victoria, great job on adding the account creation page. 🎉You've mentioned a POST request '/users' endpoint, by this endpoint do you mean for registering/creating a user? If so, the endpoint for this is already implemented. If you send a POST request to the following endpoint:
> 
> /api/v1/auth/signup
> 
> with a request body of: { "name": "victoria", "email": "[victoria@email.com](mailto:victoria@email.com)", "password": "aStrongPassword123" }
> 
> A new user with the given detail will be created and signed in.

Thank you, Mo! I've updated the code.

File: src/components/AccountCreation/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreation.tsx]: nice!

File: src/App.tsx
Comment (dargaCode) [inline src/App.tsx]: fyi I think we might have to restructure this a little bit later on when we get the cookies and stuff working, but I'm not yet sure how

File: src/components/AccountCreation/AccountCreationContainer.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreationContainer.tsx:62]: great!

File: src/components/AuthenticatedContent/AuthenticatedContent.tsx
Comment (dargaCode) [inline src/components/AuthenticatedContent/AuthenticatedContent.tsx:6]:
```suggestion
import { Link } from "react-router-dom";

```

Comment (dargaCode) [reply src/components/AuthenticatedContent/AuthenticatedContent.tsx:6]: (see below)

File: src/components/AuthenticatedContent/AuthenticatedContent.tsx
Comment (dargaCode) [inline src/components/AuthenticatedContent/AuthenticatedContent.tsx]: When working inside React and using React Router Dom to create Frontend Routes, we should use the special `<Link>` component instead of modifying the `location` or using a regular `<a>` tag to move around within the site. (We can still use `<a>` as normal to link to addresses outside of our app)

This is because we want to bypass all of the initial file loading that that browsers do when loading a url for the first time in the session. Single Page Apps already have all the initial data we need from the page we're currently on, and if we need anything extra we'll fetch it manually from the API

```suggestion
      <Link to={"/create-group"}>Create Group</Link>
```

File: src/components/AccountCreation/AccountCreation.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreation.tsx]: This is a clever solution and does work to create the line breaks!

Since a `<div>` is meant to always contain something, we should wrap each of these `<div>`s around the existing `<label>`s as their children:

```
<div>
  <label>
    ...
  </label>
</div>
```

File: src/components/AccountCreation/AccountCreationContainer.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreationContainer.tsx:34]: Now you've really created the account, so you can move the `setIsAuthenticated` in here instead of at the beginning of the handler

File: src/components/AccountCreation/AccountCreationContainer.tsx
Comment (dargaCode) [inline src/components/AccountCreation/AccountCreationContainer.tsx]: No longer need to simulate since you're making the real request now!

Please move this setter and the console.log down into the `Response.ok` conditional below. 

You should also be able to use the real `responseData` values when logging the `name` and `email` strings 🙌

**Review by dargaCode — CHANGES_REQUESTED** [review]
HI Victoria, great job with this! 

There are a couple more small changes to make here, and then all that's left will be merging in the most recent changes from `origin/dev` and resolving the merge conflict that it will cause 💪

**Comment by dargaCode** [comment]
ps Please also remove the extra `console.log()`s you're using for debugging! It's cool to keep the one for the login though since that doesn't have the real functionality visible yet

File: src/App.tsx
Comment (Victoria240) [reply src/App.tsx]: Erm, I don't know how cookies will affect the code structure. Could you please explain what you mean?

File: src/components/AuthenticatedContent/AuthenticatedContent.tsx
Comment (Victoria240) [reply src/components/AuthenticatedContent/AuthenticatedContent.tsx]: Thanks for the explanation! In this case, though, I want a button element in the UI and not a link. Is using a `<Link>` component the only fix?

File: src/App.tsx
Comment (dargaCode) [inline src/App.tsx]: very nice! please remove the `{" "}` at the end of the line though!

File: src/App.tsx
Comment (dargaCode) [reply src/App.tsx]: I have no idea how the cookies will affect things either, we'll have to figure that out this week after we learn how to use auth cookies 🤠

File: src/components/AuthenticatedContent/AuthenticatedContent.tsx
Comment (dargaCode) [reply src/components/AuthenticatedContent/AuthenticatedContent.tsx]: > I want a button element in the UI and not a link

Since this is a link to another route in the app maybe the `<Link>` would be more accurate in terms of functionality, but if you really want a `<button>` visually, you could always put one as a child of the `<Link>` tag. (Then when you click the `<button>` "Event Bubbling" will mean that you're also clicking the `<Link>`

Otherwise if you just want the link to look and feel like a button I think it might be more common to get that result via css via making the link have a background, border, and hover changes: https://codepen.io/dargaCode/full/LYqvJRP

File: src/components/AccountCreation/AccountCreation.tsx
Comment (dargaCode) [reply src/components/AccountCreation/AccountCreation.tsx]: Looks like the self-closing `<div />` are still there, unless maybe you've got local changes you forgot to push

**Review by dargaCode — APPROVED** [review]
Awesome! Please fix these last couple of unresolved issues and then you'll be good to go!

However since we just merged Annie's commit you'll get another chance to practice Merge Conflicts before you can merge it 🫠

File: src/App.tsx
Comment (dargaCode) [reply src/App.tsx]: @Victoria240 I got some advice from @angelkbrown and she gave me some topics to look into. I'll let you know when I have a clear task I can write up for how to do this and we can pair program if you want!

**Comment by dargaCode** [comment]
yayayaya great job!

**Pull request merged**
